dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java

```diff
@@ -0,0 +1,488 @@
+package com.dat3m.dartagnan.encoding;
+
+import com.dat3m.dartagnan.expression.*;
+import com.dat3m.dartagnan.expression.op.COpBin;
+import com.dat3m.dartagnan.expression.processing.ExpressionVisitor;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.event.core.Event;
+import com.dat3m.dartagnan.program.memory.MemoryObject;
+import org.sosy_lab.java_smt.api.*;
+import org.sosy_lab.java_smt.api.NumeralFormula.IntegerFormula;
+
+import static com.dat3m.dartagnan.GlobalSettings.getArchPrecision;
+import static java.util.Arrays.asList;
+
+class ExpressionEncoder implements ExpressionVisitor<Formula> {
+
+    private final FormulaManager formulaManager;
+    private final BooleanFormulaManager booleanFormulaManager;
+    private final Event event;
+
+    public ExpressionEncoder(FormulaManager formulaManager, Event event) {
+        this.formulaManager = formulaManager;
+        this.booleanFormulaManager = formulaManager.getBooleanFormulaManager();
+        this.event = event;
+    }
```

Comment:
I somewhat dislike that this Encoder works differently than the other one's, namely, that it is *used by* the `EncodingContext` while the other Encoders *use* the `EncodingContext`.

The fact that this Encoder is package-private remedies the issue though, so I'm fine with that for now.
(Feel free to mark this comment as resolved)